File: libs/shared-kernel/src/i18n.ts
import { z } from 'zod';

export type SupportedLocale = 'vi' | 'en' | 'ja' | 'ko';

export const SUPPORTED_LOCALES: SupportedLocale[] = ['vi', 'en', 'ja', 'ko'];
export const DEFAULT_LOCALE: SupportedLocale = 'vi';

export type LocalizedText = Partial<Record<SupportedLocale, string>> & { vi: string };

export const localizedTextSchema = z
  .object({
    vi: z.string().trim().min(1),
    en: z.string().trim().min(1).optional(),
    ja: z.string().trim().min(1).optional(),
    ko: z.string().trim().min(1).optional(),
  })
  .strict();

export function isSupportedLocale(value: string): value is SupportedLocale {
  return SUPPORTED_LOCALES.includes(value as SupportedLocale);
}

export function resolveLocalizedText(
  value: LocalizedText,
  requestedLocale: SupportedLocale,
  fallback: SupportedLocale = DEFAULT_LOCALE,
): string {
  const requested = value[requestedLocale];
  if (requested) return requested;

  const fallbackValue = value[fallback];
  if (fallbackValue) return fallbackValue;

  const firstAvailable = Object.values(value).find((text) => text);
  return firstAvailable ?? '';
}

File: libs/shared-kernel-go/i18n.go
package sharedkernel

type SupportedLocale string

const (
	LocaleVI SupportedLocale = "vi"
	LocaleEN SupportedLocale = "en"
	LocaleJA SupportedLocale = "ja"
	LocaleKO SupportedLocale = "ko"
)

var SupportedLocales = []SupportedLocale{LocaleVI, LocaleEN, LocaleJA, LocaleKO}

const DefaultLocale = LocaleVI

type LocalizedText map[string]string

func ResolveLocalizedText(value LocalizedText, requested SupportedLocale, fallback SupportedLocale) string {
	if text := value[string(requested)]; text != "" {
		return text
	}
	if fallback == "" {
		fallback = DefaultLocale
	}
	if text := value[string(fallback)]; text != "" {
		return text
	}
	for _, text := range value {
		if text != "" {
			return text
		}
	}
	return ""
}
